doc/developer/design/20220404_metadata_storage.md

```diff
@@ -0,0 +1,296 @@
+migrate to CockroachDB, which is largely PostgreSQL compatible, anad has
+effectively no limits on horizontal scalability.
+
+We also have one hard fix, which is to migrate to FoundationDB. The `Stash`
+API should facilitate this, if necessary.
+
+I further propose that, for development velocity, we evaluate *only* Aurora
+PostgreSQL for now. If we verify that `Stash` observes the required single-digit
+millisecond commit latencies when backed by Aurora PostgreSQL in production, I
+don't think we need to run PoCs with CockroachDB or FoundationDB. Aurora
+PostgreSQL will scale sufficiently well for several months, if not years; that
+will give us plenty of time down the road to run PoCs with CockroachDB or
+FoundationDB if it becomes necessary.
+
+## Alternatives considered
+
+### Cross-environment reporting queries
+
+The ability to run cross-environment reporting queries has been suggested as
+a hard requirement for the metadata database. There are three known use cases
+for this:
+
+  1. Exposing a global view of the region to the ops team.
+  2. Providing business intelligence via Fivetran-managed CDC to Snowflake.
+  3. Powering reporting jobs like billing.
+
+Essentially, instead of having separate databases for each environment...
+
+```sql
+CREATE DATABASE environment_1;
+CREATE TABLE environment_1.clusters (id int, spec text);
+CREATE TABLE environment_1.view s(id int, sql text);
+
+CREATE DATABASE environment_2;
+CREATE TABLE environment_2.clusters (id int, spec text);
+CREATE TABLE environment_2.views (id int, sql text);
+```
+
+...we'd have a single database where each table has an environment ID:
+
+
+```sql
+CREATE DATABASE aws_us_east_1;
+CREATE TABLE aws_us_east_1.clusters (environment_id int, id int, spec text);
+CREATE TABLE aws_us_east_1.views (environment_id, id int, sql text);
+```
+
+This allows:
+
+  1. The ops team to run queries like `SELECT * FROM clusters` to see all
+     clusters across all customer environments.
+  2. Configuration of a single Fivetran sync job to export all data in this
+     database to Snowflake.
+  3. A single billing job in each region to query the database to determine
+     what clusters were created and deleted across all customer environments.
+
+Unfortunately there are several downsides to this approach:
+
+  * Schema changes would not be isolated to each environment. Every change to the
+    schema would require a two step process: first the new column would need to
+    be added to PostgreSQL; only *then* could environments be upgraded to the
+    new `materialized`. Rollbacks are similarly difficult: it might be
+    impossible to roll back an environment if a backwards-incompatible schema
+    change has been deployed to the region. The logistics of these schema
+    changes would be quite delicate, as the definition of the schema would
+    live in the MaterializeInc/materialize repository but the application of the
+    migration would be managed by MaterializeInc/cloud.
```

Comment:
I'm an amateur here, but it seems like this depends a bit on how you want to do schema changes. If you use `ALTER` you do require there be only one instance of whatever is holding the data, and folks need to be sync'd up. Alternately, you could create new tables with new columns, and migrate rows from the old to the new (and potentially back again, for rollbacks). I'm not sure that this doesn't have other problems, but it is afaiu the way we are thinking of doing schema migrations internal to MZ, so if it is really bad that would be good to know for other reasons! :D

Reply:
I think this section is pointing out that we can do schema upgrade more granularly with the proposal above? Because we can alter the tables for a single environment at a time, we can do schema change canaries for test env's, and limit the blast radius of complex upgrades and rollbacks